src/app/api/core/utils/withErrorHandler.ts

```diff
@@ -0,0 +1,50 @@
+import { NextRequest, NextResponse } from 'next/server'
+import { ZodError, ZodIssue } from 'zod'
+import { CopilotApiError } from '@/types/CopilotApiError'
+import APIError from '../exceptions/api'
+
+type RequestHandler = (req: NextRequest, params: any) => Promise<NextResponse>
+
+/**
+ * Wraps a given request handler with a global error handler to standardize response structure
+ * in case of failures. Catches exceptions thrown from the handler, and returns a formatted error response.
+ *
+ * @param {RequestHandler} handler - The request handler to wrap.
+ * @returns {RequestHandler} The new handler that includes error handling logic.
+ * @example
+ * const safeHandler = withErrorHandler(async (req: NextRequest) => {
+ *   // your request handling logic
+ *   if (errorCondition) {
+ *     throw new Error("Oh no!")}
+ *   return NextResponse.next();
+ * });
+ *
+ * @throws {ZodError} Captures and handles validation errors and responds with status 400 and the issue detail.
+ * @throws {CopilotApiError} Captures and handles CopilotAPI errors, uses the error status, and message if available.
+ * @throws {APIError} Captures and handles APIError
+ */
+export const withErrorHandler = (handler: RequestHandler): RequestHandler => {
+  return async (req: NextRequest, params: any) => {
+    try {
+      return await handler(req, params)
+    } catch (error) {
+      console.error(error)
+
+      let status = 500
+      let message: string | ZodIssue[] = 'Something went wrong'
+
+      if (error instanceof ZodError) {
```

Comment:
zod error can be 422 as well?

Reply:
@sazanrjb this is just the default value for status. I have changed it to `422` in `ZodError` below now.